### PR TITLE
feat(earn): prepare transactions and use on confirmation screen

### PR DESCRIPTION
**Pull request by MuckT**
### Description

Handles the preparation of all earn transactions via a shared hook that's used on both `EarnEnterAmount.tsx` & `EarnConfirmationScreen.tsx`. 

### Test plan

- [x] Tested locally on iOS
- [x] Tested locally on Andorid
- [x] Unit Tests updated

### Related issues

- Part of ACT-1389

### Backwards compatibility

Yes

### Network scalability

N/A

### CODE REVIEW

File: src/earn/EarnConfirmationScreen.tsx [hunk 1]
```diff
@@ -74,20 +78,50 @@ export default function EarnConfirmationScreen({ route }: Props) {
...
-  } = usePrepareWithdrawAndClaimTransactions({
-    amount: new BigNumber(withdrawAmountInDepositToken).dividedBy(pool.pricePerShare[0]).toString(),
-    pool,
-    walletAddress,
-    feeCurrencies,
-    hooksApiUrl,
-    rewardsPositions,
-    useMax,
-  })
+    // @ts-expect-error prepareTransactionsResult & swapTransaction are only set when mode === 'deposit' or 'swap-deposit'
```
Comment (jh2oman) [inline src/earn/EarnConfirmationScreen.tsx:82]: I don't understand these two lines. 
1. Seems like a red flag to do a ts-ignore here. if the return values of the different modes are that different then should they maybe be changed at the prepare transaction level? 
2. swapTransaction doesn't even seem to be used in the code.

File: src/earn/EarnConfirmationScreen.tsx [hunk 2]
```diff
@@ -74,20 +78,50 @@ export default function EarnConfirmationScreen({ route }: Props) {
...
+
+    return refreshPreparedTransactions({
+      amount: new BigNumber(amount).dividedBy(pool.pricePerShare[0]).toString(),
+      token,
+      walletAddress,
+      feeCurrencies,
+      pool,
+      hooksApiUrl,
+      // Mode === shortcutId, except for exit which is withdraw
+      shortcutId: mode === 'exit' ? 'withdraw' : mode,
```
Comment (jh2oman) [inline src/earn/EarnConfirmationScreen.tsx:108]: So exit maps to 'prepareWithdrawAndClaimTransactions' which doesn't appear to have shortcutId as a parameter. So why is this needed?

File: src/earn/EarnEnterAmount.tsx [hunk 3]
```diff
@@ -144,6 +144,7 @@ function EarnEnterAmount({ route }: Props) {
   }
 
   const {
+    // @ts-expect-error prepareTransactionsResult & swapTransaction are only set when mode === 'deposit' or 'swap-deposit'
```
Comment (jh2oman) [inline src/earn/EarnEnterAmount.tsx:147]: Again, the ts-expects-error is a big red flag. The underlying types should probably be changed or we just need separate functions

File: src/earn/prepareTransactions.ts [hunk 4]
```diff
@@ -166,52 +173,79 @@ export async function prepareWithdrawTransactions({
...
+  walletAddress: Address
+  feeCurrencies: TokenBalance[]
+  hooksApiUrl: string
+  rewardsPositions: Position[]
+}) {
+  const { appId, networkId } = pool
 
+  // Prepare the claim transactions if there are rewards positions
+  const claimTransactions = rewardsPositions
+    ? await Promise.all(
```
Comment (jh2oman) [inline src/earn/prepareTransactions.ts:224]: this checks if `rewardsPositions` is falsy, but can it ever be falsy since its guarenteed to be an array?

File: src/earn/EarnConfirmationScreen.tsx (hunk 1 above)
Comment (MuckT) [reply src/earn/EarnConfirmationScreen.tsx:82]: This @ts-expect-error is temporary; I'll remove it before moving the PR out of draft and resolving the type issues. While `swapTransaction` isn't currently used on this page there's [a bug](https://linear.app/valora/issue/ACT-1398) with the bottom sheet implementation that will likely require `swapTransaction` to be handled from a confirmation screen.

File: src/earn/prepareTransactions.ts (hunk 4 above)
Comment (MuckT) [reply src/earn/prepareTransactions.ts]: Fixed in e9f2c075215324c83c03db3c107e9b1a62df6523.

File: src/earn/EarnConfirmationScreen.tsx (hunk 2 above)
Comment (MuckT) [reply src/earn/EarnConfirmationScreen.tsx]: Fixed in 98a8bd00d6a0d1ecef892ad8d7bdc34d39e86b63.

File: src/earn/EarnConfirmationScreen.tsx (hunk 1 above)
Comment (MuckT) [reply src/earn/EarnConfirmationScreen.tsx]: This `// @ts-expect-error` is no longer present but the one on `EarnEnterAmount.tsx` is still present; I'll see how this can be removed.

File: src/earn/prepareTransactions.ts [hunk 5]
```diff
@@ -112,106 +114,130 @@ export async function prepareWithdrawAndClaimTransactions({
...
+    rewardsPositions.map(async (position): Promise<RawShortcutTransaction[]> => {
+      const { transactions }: { transactions?: RawShortcutTransaction[] } =
+        await triggerShortcutRequest(hooksApiUrl, {
+          address: walletAddress,
+          appId,
+          networkId,
+          shortcutId: 'claim-rewards',
+          ...position.shortcutTriggerArgs?.['claim-rewards'],
+        })
+      return transactions ?? [] // Default to an empty array if rewardsPositions is undefined
```
Comment (satish-ravi) [inline src/earn/prepareTransactions.ts:227]: little confused by this comment, if rewardPositions is undefined, this fn would never be executed right?

File: src/earn/hooks.ts [hunk 6]
```diff
@@ -94,3 +90,52 @@ export function useDepositEntrypointInfo({
...
+    switch (mode) {
+      case 'deposit':
+      case 'swap-deposit':
+        return prepareDepositTransactions
+      case 'withdraw':
+        return prepareWithdrawTransactions
+      case 'claim-rewards':
+        return prepareClaimTransactions
+      case 'exit':
+        return prepareWithdrawAndClaimTransactions
```
Comment (satish-ravi) [inline src/earn/hooks.ts:108]: having all the modes here is nice, but I wonder if it does more harm that good. This hook I believe is mainly for enter amount screen which needs the functionality to refresh transactions as the user input changes, but for claim and exit do we even need it (related to my other comment)? Wonder if this would be simpler to just wrap the transaction preparation in a `useAsync` like how it's done currently. You also wouldn't need the args hack you have below.

File: src/earn/EarnConfirmationScreen.test.tsx [hunk 7]
```diff
@@ -187,15 +183,6 @@ describe('EarnConfirmationScreen', () => {
     expect(getByTestId('EarnConfirmation/GasFeeFiatAmount')).toHaveTextContent('₱119.70')
     expect(queryByTestId('EarnConfirmation/GasSubsidized')).toBeFalsy()
     expect(getByTestId('EarnConfirmationScreen/CTA')).toBeEnabled()
-    // TODO(act-1389): update this test to make sure that reward positions are not included in partial withdrawals.
```
Comment (jh2oman) [inline src/earn/EarnConfirmationScreen.test.tsx:190]: Can you add this check the AC refers to?

File: src/earn/EarnConfirmationScreen.tsx [hunk 8]
```diff
@@ -77,19 +77,18 @@ export default function EarnConfirmationScreen({ route }: Props) {
...
-  } = usePrepareWithdrawAndClaimTransactions({
+  } = usePrepareClaimExitAndWithdrawTransactions(mode, {
     amount: withdrawAmountInDepositToken.dividedBy(pool.pricePerShare[0]).toString(),
     pool,
     walletAddress,
     feeCurrencies,
     hooksApiUrl,
     rewardsPositions,
-    useMax,
+    useMax: !!useMax,
```
Comment (jh2oman) [inline src/earn/EarnConfirmationScreen.tsx:91]: I didn't notice that useMax was boolean | undefined in a previous pr. Can you just make it a boolean type? That would be cleaner.

File: src/earn/hooks.ts [hunk 9]
```diff
@@ -94,3 +92,68 @@ export function useDepositEntrypointInfo({
   }, [asyncExchanges.result])
   return { hasDepositToken, hasTokensOnSameNetwork, hasTokensOnOtherNetworks, canCashIn, exchanges }
 }
+
+export function usePrepareDepositAndWithdrawTransactions(
+  mode: Extract<EarnActiveMode, 'deposit' | 'withdraw' | 'swap-deposit'>
```
Comment (jh2oman) [inline src/earn/hooks.ts:97]: When i read these functions, I think to myself
> Why are deposit, withdraw, and swap-deposit in one function and claim-rewards, exit, and withdraw in another function?
> Why are withdraw in both functions and why does one call prepareWithdrawTransactionsWithSwap and the other calls  prepareWithdrawTransactions.

The short answer to both of these, i believe, is because the two functions are used in two different screens and the grouping is purely based on UX choices which is very hard to determine by just looking at these functions. Can they be re-named to be more closely associated with the screens they are used in and maybe comments added to help explain why these two functions exist?

File: src/earn/hooks.ts [hunk 10]
```diff
@@ -94,3 +92,68 @@ export function useDepositEntrypointInfo({
...
+  return {
+    prepareTransactionsResult: prepareTransactions.result,
+    refreshPreparedTransactions: prepareTransactions.execute,
+    clearPreparedTransactions: prepareTransactions.reset,
+    prepareTransactionError: prepareTransactions.error,
+    isPreparingTransactions: prepareTransactions.loading,
+  }
+}
+
+export function usePrepareClaimExitAndWithdrawTransactions(
```
Comment (jophish) [inline src/earn/hooks.ts:132]: Not a huge deal, but it's slightly strange to me that one of these functions (this one, `usePrepareClaimExitAndWithdrawTransactions`) actually performs the call on the TX preparer, and the other (`usePrepareDepositAndWithdrawTransactions`) just returns a callback fn + execution state. I wonder if the naming of `usePrepareDepositAndWithdrawTransactions` could be updated to reflect this, e.g. `usePrepareDepositAndWithdrawTransactionsCallback`.

File: src/earn/prepareTransactions.ts [hunk 11]
```diff
@@ -118,100 +123,169 @@ export async function prepareWithdrawAndClaimTransactions({
       ],
       ...shortcutTriggerArgs?.withdraw,
     })
+  Logger.debug(TAG, 'prepareWithdrawTransactions', {
```
Comment (jophish) [inline src/earn/prepareTransactions.ts:126]: nit: Update this fn name in the log to match the actual fn name

File: src/earn/prepareTransactions.ts [hunk 12]
```diff
@@ -118,100 +123,169 @@ export async function prepareWithdrawAndClaimTransactions({
...
+  // Is there a better way to do this?
+  // Pass a param to a shared prepareTransactions function to indicate that the transaction should also return swapTransaction:
+  return {
+    prepareTransactionsResult: await prepareTransactions({
+      feeCurrencies,
+      baseTransactions: rawShortcutTransactionsToTransactionRequests(withdrawTransactions),
+      isGasSubsidized: isGasSubsidizedForNetwork(networkId),
+      origin: 'earn-withdraw',
+    }),
+    swapTransaction: undefined,
```
Comment (jophish) [inline src/earn/prepareTransactions.ts:140]: Does this `swapTransaction: undefined` actually need to be here?

File: src/earn/prepareTransactions.ts [hunk 13]
```diff
@@ -118,100 +123,169 @@ export async function prepareWithdrawAndClaimTransactions({
...
+  }
+}
+
+// Used on the EarnConfirmationScreen.tsx
+export async function prepareWithdrawAndClaimTransactions({
```
Comment (jophish) [inline src/earn/prepareTransactions.ts:145]: Is it possible for `prepareWithdrawAndClaimTransactions` to re-use both `prepareWithdrawTransactions` and `prepareClaimTransactions`?

File: src/earn/EarnConfirmationScreen.tsx (hunk 8 above)
Comment (MuckT) [reply src/earn/EarnConfirmationScreen.tsx:91]: Fixed in e2d92f15d1a4aeab5cdc6bdac95792d010558075!

File: src/earn/prepareTransactions.ts (hunk 12 above)
Comment (MuckT) [reply src/earn/prepareTransactions.ts:140]: It is otherwise we get a type error on `usePrepareDepositAndWithdrawTransactions` in `EarnEnterAmount.tsx`: 
```
Property 'swapTransaction' does not exist on type '{ prepareTransactionsResult: PreparedTransactionsResult; } | {}'
```

File: src/earn/prepareTransactions.ts (hunk 11 above)
Comment (MuckT) [reply src/earn/prepareTransactions.ts]: Fixed in 501c1c34409b34d4835f919c48cb863eb29ae49d!

File: src/earn/hooks.ts (hunk 9 above)
Comment (MuckT) [reply src/earn/hooks.ts:98]: @jh2oman That's correct, the grouping is based on which screen they are used on, and withdraw is used on both the `EarnEnterAmount.tsx` & `EarnConfrimationScreen.tsx`. I've updated the names and added a short comment on why these separate hooks are needed in 457a5945c8df94df36661f7b934ab0b2c17a4ce1.

File: src/earn/hooks.ts [hunk 14]
```diff
@@ -94,3 +92,68 @@ export function useDepositEntrypointInfo({
   }, [asyncExchanges.result])
   return { hasDepositToken, hasTokensOnSameNetwork, hasTokensOnOtherNetworks, canCashIn, exchanges }
 }
+
+export function usePrepareDepositAndWithdrawTransactions(
+  mode: Extract<EarnActiveMode, 'deposit' | 'withdraw' | 'swap-deposit'>
+) {
+  const getTransactionFunction = () => {
+    switch (mode) {
```
[the remark this answers is not in the document]
Comment (MuckT) [reply src/earn/hooks.ts]: I opted to go for a ternary for simplicity in 34420b9cd750c54a95c9ecfc118b1c87e9ad817c.

File: src/earn/hooks.ts (hunk 10 above)
Comment (MuckT) [reply src/earn/hooks.ts]: Renamed in 3de16cdb20bb644b0541382d5e4b517f4fede2ed.

File: src/earn/prepareTransactions.ts (hunk 13 above)
Comment (MuckT) [reply src/earn/prepareTransactions.ts:145]: Something the code below works, but I get a return type error: `Property 'transactions' does not exist on type 'PreparedTransactionsResult'.` As this is a pre-existing function I think it's fine to leave as is.

```TypeScript
export async function prepareWithdrawAndClaimTransactions({
  walletAddress,
  feeCurrencies,
  pool,
  hooksApiUrl,
  rewardsPositions,
}: {
  walletAddress: Address
  feeCurrencies: TokenBalance[]
  pool: EarnPosition
  hooksApiUrl: string
  rewardsPositions: Position[]
}) {
  const { transactions: withdrawTransactions } = await prepareWithdrawTransactions({
    amount: pool.balance,
    walletAddress,
    feeCurrencies,
    pool,
    hooksApiUrl,
    useMax: true,
  })

  const { transactions: claimTransaction } = await prepareClaimTransactions({
    walletAddress,
    feeCurrencies,
    pool,
    hooksApiUrl,
    rewardsPositions,
  })

  Logger.debug(TAG, 'prepareWithdrawAndClaimTransactions', {
    withdrawTransactions,
    claimTransaction,
    pool,
  })

  const combinedTransactions = [...withdrawTransactions, ...claimTransaction]

  return prepareTransactions({
    feeCurrencies,
    baseTransactions: rawShortcutTransactionsToTransactionRequests(combinedTransactions),
    isGasSubsidized: isGasSubsidizedForNetwork(pool.networkId),
    origin: 'earn-claim-rewards',
  })
}
```

File: src/earn/EarnConfirmationScreen.test.tsx (hunk 7 above)
Comment (MuckT) [reply src/earn/EarnConfirmationScreen.test.tsx:190]: Small one added in 502a676148bbc2a0e3c217baa195907330415dd2.

File: src/earn/prepareTransactions.ts [hunk 15]
```diff
@@ -93,16 +92,22 @@ export async function prepareDepositTransactions({
   }
 }
 
-export async function prepareWithdrawAndClaimTransactions({
-  pool,
+export async function prepareWithdrawTransactionsWithSwap({
```
Comment (MuckT) [inline src/earn/prepareTransactions.ts:95]: @satish-ravi Tried a few ways to make this work without having this additional function but came up short.

Comment (satish-ravi) [reply src/earn/prepareTransactions.ts:95]: you can still avoid duplicating the fn by calling `prepareWithdrawTransactions` from this function and adjusting the response type right? This is exactly what we did in our pairing session except we did it with an inline function in the hook instead of an exported function here

Comment (MuckT) [reply src/earn/prepareTransactions.ts:96]: Was able to make this change in 8cffb5b8d2fcf4f24c47502d866e4a59c6c8a8f8 and it is a lot cleaner. I tried to do something similar with `prepareWithdrawAndClaimTransactions`, but ran into issues merging the `prepareTransactionsResult` objects.

File: src/earn/EarnConfirmationScreen.test.tsx [hunk 16]
```diff
@@ -138,22 +143,15 @@ describe('EarnConfirmationScreen', () => {
     await waitFor(() => {
       expect(queryByTestId('EarnConfirmation/GasLoading')).toBeFalsy()
     })
+
     expect(getByTestId('EarnConfirmation/GasFeeCryptoAmount')).toHaveTextContent('0.06 ETH')
     expect(getByTestId('EarnConfirmation/GasFeeFiatAmount')).toHaveTextContent('₱119.70')
     expect(queryByTestId('EarnConfirmation/GasSubsidized')).toBeFalsy()
     expect(getByTestId('EarnConfirmationScreen/CTA')).toBeEnabled()
-    expect(prepareWithdrawAndClaimTransactions).toHaveBeenCalledWith({
```
Comment (jh2oman) [inline src/earn/EarnConfirmationScreen.test.tsx:145]: Can you verify that the appropriate prepareTransaction function is called with the expected arguments?

File: src/earn/prepareTransactions.ts [hunk 17]
```diff
@@ -112,106 +152,125 @@ export async function prepareWithdrawAndClaimTransactions({
...
     claimTransactions,
     pool,
   })
+
   return prepareTransactions({
     feeCurrencies,
     baseTransactions: rawShortcutTransactionsToTransactionRequests([
       ...withdrawTransactions,
-      ..._.flatten(claimTransactions),
+      ...(dataProps.withdrawalIncludesClaim ? [] : claimTransactions.flat()),
```
Comment (jh2oman) [inline src/earn/prepareTransactions.ts:185]: Do we even need to call the `'claim-rewards'` shortcut if withdrawalIncludesClaim === true ?

Comment (jh2oman) [reply src/earn/prepareTransactions.ts:185]: Also please add a test for this added logic

Comment (MuckT) [reply src/earn/prepareTransactions.ts:185]: Calling `claim-rewards` if withdrawalIncludesClaim === true allows the users to claim-rewards without withdrawing.

Comment (MuckT) [reply src/earn/prepareTransactions.ts]: That's a good point; we don't need to call `'claim-rewards'` here. Code change a65343a36aa1609ff4ae4008a1936b92cc361f4e and test 47787a8b084e759a52ca1d9d2bd103cbab2f43ac.

File: src/earn/EarnConfirmationScreen.test.tsx (hunk 16 above)
Comment (MuckT) [reply src/earn/EarnConfirmationScreen.test.tsx]: Added in 26316f89a7e4947617fa7a1a01659e44823bb167!

File: src/earn/hooks.ts [hunk 18]
```diff
@@ -94,3 +92,60 @@ export function useDepositEntrypointInfo({
...
+    prepareTransactionsResult: prepareTransactions.result,
+    refreshPreparedTransactions: prepareTransactions.execute,
+    clearPreparedTransactions: prepareTransactions.reset,
+    prepareTransactionError: prepareTransactions.error,
+    isPreparingTransactions: prepareTransactions.loading,
+  }
+}
+
+// Used on the EarnConfirmationScreen.tsx called once on load
+export function usePrepareEnterConfirmationScreenTransactions(
```
Comment (jh2oman) [inline src/earn/hooks.ts:124]: nit: i think you mean to name it usePrepareConfirmationScreenTransactions

**Review by jh2oman — APPROVED** [review]

File: src/earn/hooks.ts (hunk 18 above)
Comment (MuckT) [reply src/earn/hooks.ts]: Renamed to `usePrepareEarnConfirmationScreenTransactions` in d1558f3e935968571b4b2e54528fb0f1de204535.

**Review by jophish — APPROVED** [review]
